t/t1091-sparse-checkout-builtin.sh

```diff
@@ -0,0 +1,216 @@
+#!/bin/sh
+
+test_description='sparse checkout builtin tests'
+
+. ./test-lib.sh
+
+test_expect_success 'setup' '
+	git init repo &&
+	(
+		cd repo &&
+		echo "initial" >a &&
+		mkdir folder1 folder2 deep &&
+		mkdir deep/deeper1 deep/deeper2 &&
+		mkdir deep/deeper1/deepest &&
+		cp a folder1 &&
+		cp a folder2 &&
+		cp a deep &&
+		cp a deep/deeper1 &&
+		cp a deep/deeper2 &&
+		cp a deep/deeper1/deepest &&
+		git add . &&
+		git commit -m "initial commit"
+	)
+'
+
+test_expect_success 'git sparse-checkout list (empty)' '
+	git -C repo sparse-checkout list >list 2>err &&
+	test_line_count = 0 list &&
+	test_i18ngrep "this worktree is not sparse (sparse-checkout file may not exist)" err
+'
+
+test_expect_success 'git sparse-checkout list (populated)' '
+	test_when_finished rm -f repo/.git/info/sparse-checkout &&
+	cat >repo/.git/info/sparse-checkout <<-EOF &&
+		/folder1/*
+		/deep/
+		**/a
+		!*bin*
+	EOF
+	git -C repo sparse-checkout list >list &&
+	cat >expect <<-EOF &&
+		/folder1/*
+		/deep/
+		**/a
+		!*bin*
+	EOF
+	test_cmp expect list
+'
+
+test_expect_success 'git sparse-checkout init' '
+	git -C repo sparse-checkout init &&
+	cat >expect <<-EOF &&
+		/*
+		!/*/*
+	EOF
+	test_cmp expect repo/.git/info/sparse-checkout &&
+	git -C repo config --list >config &&
+	test_i18ngrep "core.sparsecheckout=true" config &&
+	ls repo >dir  &&
+	echo a >expect &&
+	test_cmp expect dir
+'
+
+test_expect_success 'git sparse-checkout list after init' '
+	git -C repo sparse-checkout list >actual &&
+	cat >expect <<-EOF &&
+		/*
+		!/*/*
+	EOF
+	test_cmp expect actual
+'
+
+test_expect_success 'init with existing sparse-checkout' '
+	echo "/folder1/*" >> repo/.git/info/sparse-checkout &&
+	git -C repo sparse-checkout init &&
+	cat >expect <<-EOF &&
+		/*
+		!/*/*
+		/folder1/*
+	EOF
+	test_cmp expect repo/.git/info/sparse-checkout &&
+	ls repo >dir  &&
+	cat >expect <<-EOF &&
+		a
+		folder1
+	EOF
+	test_cmp expect dir
+'
+
+test_expect_success 'clone --sparse' '
+	git clone --sparse repo clone &&
+	git -C clone sparse-checkout list >actual &&
+	cat >expect <<-EOF &&
+		/*
+		!/*/*
+	EOF
+	test_cmp expect actual &&
+	ls clone >dir &&
+	echo a >expect &&
+	test_cmp expect dir
+'
+
+test_expect_success 'set sparse-checkout using builtin' '
+	git -C repo sparse-checkout set "/*" "!/*/" "/folder1/" "/folder2/" &&
+	cat >expect <<-EOF &&
+		/*
+		!/*/
+		/folder1/
+		/folder2/
```

Comment:
I fear that we need something slightly uglier here, as this test [fails on Windows](https://dev.azure.com/gitgitgadget/git/_build/results?buildId=16385&view=ms.vss-test-web.build-test-results-tab&runId=44608&resultId=100343&paneView=debug) with this:

```
+++ diff -u expect actual
--- expect	2019-09-18 02:45:52.220988300 +0000
+++ actual	2019-09-18 02:45:52.316987900 +0000
@@ -1,4 +1,4 @@
 /*
 !/*/
-/folder1/
-/folder2/
+D:/a/1/s/git-sdk-64-minimal/folder1/
+D:/a/1/s/git-sdk-64-minimal/folder2/
error: last command exited with $?=1
```

The easiest method might look like this:

```
root_prefix= &&
{ test_have_prereq !MINGW || root_prefix="$(cygpath -wm /)"; } &&
[...]
    $root_prefix/folder1/
    $root_prefix/folder2/
```

Reply:
Thanks, @dscho. I noticed that failure, but didn't see that the interpretation of "root" was being output strangely. I think it worked with `/*` at the end of the patterns, but I'll double-check.

Feedback from the list made me try it, but somehow this type of pattern doesn't work nicely on Windows.

Reply:
Thanks, @dscho. I noticed that failure, but didn't see that the interpretation of "root" was being output strangely. I think it worked with /* at the end of the patterns, but I'll double-check.

Feedback from the list made me try it, but somehow this type of pattern doesn't work nicely on Windows.

Reply:
This is a failure in the SDK, it seems! I'm passing "/folder1/" as an argument, and the SDK shell is converting it to `C:/git-sdk-64/folder1/`. Any ideas how to not do this, @dscho?

Reply:
(It doesn't happen when I pass `/folder1/*`)

Short term, I can just stop using these root-based patterns in the "set" tests (without cones).

Reply:
This is in our [known issues](https://github.com/git-for-windows/build-extra/blob/master/ReleaseNotes.md#known-issues):

> * If you specify command-line options starting with a slash, POSIX-to-Windows path conversion will kick in converting e.g. "`/usr/bin/bash.exe`" to "`C:\Program Files\Git\usr\bin\bash.exe`". When that is not desired -- e.g. "`--upload-pack=/opt/git/bin/git-upload-pack`" or "`-L/regex/`" -- you need to set the environment variable `MSYS_NO_PATHCONV` temporarily, like so:
> 
>  > `MSYS_NO_PATHCONV=1 git blame -L/pathconv/ msys2_path_conv.cc`
>
>  Alternatively, you can double the first slash to avoid POSIX-to-Windows path conversion, e.g. "`//usr/bin/bash.exe`".